---
title: "Gviz visualization of DPP8/9 genetic variation (hg19)"
output: html_document
---

### Fill in parameters

--- Before running this Rmd document, please fill in the following parameters: gene and typeofvariant ---

```{r para}
gene <- "Dipeptidyl Peptidase 9"
typeofvariant <- "Copy Number Variation"

if(gene == "Dipeptidyl Peptidase 9"){start = 4675239 
                                    end = 4723855}
if(gene == "Dipeptidyl Peptidase 8"){start = 65734805
                                    end = 65810040}
end1 <- start-6240
end2 <- end+1380
```

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
```

```{r Rversion}
R.version$os
R.version.string
```

### Loading in GeneModel file

Make sure this R script is located in the correct subfolder where the NCBI results are.

```{r genemodel}
GeneModel <- read.csv(file="results-transcripts-UCSC-hg19.csv", header = TRUE, sep=";")
```

### Install and load Gviz Bioconductor package

```{r installation}
if (! requireNamespace("BiocManager", quietly = TRUE)){
   install.packages("BiocManager")}
if (! requireNamespace("Gviz", quietly = TRUE)) {
   BiocManager::install("Gviz")}
library(Gviz)
```

Bioconductor version:

```{r Bioconductor}
packageVersion("BiocManager")
```

### Check the version of the packages installed

```{r version}
packageVersion("Gviz")
packageVersion("S4Vectors")
packageVersion("stats4")
packageVersion("BiocGenerics")
packageVersion("parallel")
packageVersion("IRanges")
packageVersion("GenomicRanges")
packageVersion("GenomeInfoDb")
packageVersion("grid")
```

### Select the file containing the data to be visualized

```{r file}
filename <- read.csv(file=file.choose(), header = TRUE, sep = ";")
```

### Function to select the right columns from the table 

```{r function}
if (! requireNamespace("tidyr", quietly = TRUE)){
   install.packages("tidyr")}
library(tidyr)

Gviz_table <- function(filetable) {
  X <- filetable
 #################### GRCh37 coordinates in first assembly
  tablepart1 <- X[grep("^GRCh37", X$assembly1),]
  id1 <- tablepart1[1]
  if(any(names(tablepart1) == 'variation.type')){type1 <- tablepart1[3]} else {type1 <- tablepart1[5]}
  if(any(names(tablepart1) == 'variant.region.id')){regionid1 <- tablepart1[2]}
  if(any(names(tablepart1) == 'accession')){acc1 <- tablepart1[3]}
  Chr_1a <- tablepart1[["Chr"]]
  Chr_1a <- gsub("[A-Za-z]", "", Chr_1a)
  Chr_1a <- sub("", "chr", Chr_1a)
  assembly1 <- tablepart1[["assembly1"]]
  assembly1 <- gsub(".*?:", "", assembly1)
  assembly1 <- separate(data = as.data.frame(assembly1), col = assembly1, into = c("start", "end"), sep = "-")
  if(any(names(tablepart1) == 'variant.region.id')) {
    GRCh37_assembly1 <- data.frame(id1,Chr_1a,assembly1,regionid1,type1)
  } else if(any(names(tablepart1) == 'accession')) {
    GRCh37_assembly1 <- data.frame(id1,Chr_1a,assembly1,acc1,type1)
  } else {
    GRCh37_assembly1 <- data.frame(id1,Chr_1a,assembly1,type1)
  }
  names(GRCh37_assembly1)[names(GRCh37_assembly1) == "assembly1"] <- "assembly"
  names(GRCh37_assembly1)[names(GRCh37_assembly1) == "Chr_1a"] <- "chr"
  #################### GRCh37 coordinates in second assembly
  tablepart2 <- X[grep("^GRCh37", X$assembly2),]
  id2 <- tablepart2[1]
  if(any(names(tablepart2) == 'variation.type')){type2 <- tablepart2[3]} else {type2 <- tablepart2[5]}
  if(any(names(tablepart2) == 'variant.region.id')){regionid2 <- tablepart2[2]}
  if(any(names(tablepart2) == 'accession')){acc2 <- tablepart2[3]}
  Chr_1b <- tablepart2[["Chr"]]
  Chr_1b <- gsub("[A-Za-z]", "", Chr_1b)
  Chr_1b <- sub("", "chr", Chr_1b)
  assembly2 <- tablepart2[["assembly2"]]
  assembly2 <- gsub(".*?:", "", assembly2)
  assembly2 <- separate(data = as.data.frame(assembly2), col = assembly2, into = c("start", "end"), sep = "-")
  if(any(names(tablepart2) == 'variant.region.id') &
      any(names(tablepart2) == 'study.ID')){
    GRCh37_assembly2 <- data.frame(id2,Chr_1b,assembly2,regionid2,type2)
  } else if(any(names(tablepart2) == 'accession')) {
    GRCh37_assembly2 <- data.frame(id2,Chr_1b,assembly2,acc2,type2)
  } else {
    GRCh37_assembly2 <- data.frame(id2,Chr_1b,assembly2,type2)
  }
  names(GRCh37_assembly2)[names(GRCh37_assembly2) == "assembly2"] <- "assembly"
  names(GRCh37_assembly2)[names(GRCh37_assembly2) == "Chr_1b"] <- "chr"
  #################### GRCh37 coordinates in third assembly
  tablepart3 <- X[grep("^GRCh37", X$assembly3),]
  id3 <- tablepart3[1]
  if(any(names(tablepart3) == 'variation.type')){type3 <- tablepart3[3]} else {type3 <- tablepart3[5]}
  if(any(names(tablepart3) == 'variant.region.id')){regionid3 <- tablepart3[2]}
  if(any(names(tablepart3) == 'accession')){acc3 <- tablepart3[3]}
  Chr_1c <- tablepart3[["Chr"]]
  Chr_1c <- gsub("[A-Za-z]", "", Chr_1c)
  Chr_1c <- sub("", "chr", Chr_1c)
  assembly3 <- tablepart3[["assembly3"]]
  assembly3 <- gsub(".*?:", "", assembly3)
  assembly3 <- separate(data = as.data.frame(assembly3), col = assembly3, into = c("start", "end"), sep = "-")
  if(any(names(tablepart3) == 'variant.region.id') &
      any(names(tablepart3) == 'study.ID')){
    GRCh37_assembly3 <- data.frame(id3,Chr_1c,assembly3,regionid3,type3)
  } else if(any(names(tablepart3) == 'accession')) {
    GRCh37_assembly3 <- data.frame(id3,Chr_1c,assembly3,acc3,type3)
  } else {
    GRCh37_assembly3 <- data.frame(id3,Chr_1c,assembly3,type3)
  }
  names(GRCh37_assembly3)[names(GRCh37_assembly3) == "assembly3"] <- "assembly"
  names(GRCh37_assembly3)[names(GRCh37_assembly3) == "Chr_1c"] <- "chr"
  ##### Gviz table
  GRCh37_Gviz <- rbind(GRCh37_assembly1,GRCh37_assembly2,GRCh37_assembly3)
  GRCh37_Gviz <- GRCh37_Gviz[order(-GRCh37_Gviz[1]),]
}
```

### Make GRanges Object from data frame and set colors for different types

```{r GRanges}
TableObject <- Gviz_table(filename)
TableObject
names(TableObject)[names(TableObject) == "variant.call.type"] <- "type"
TableObject$start[as.numeric(TableObject$start)<end1] <- end1
TableObject$end[as.numeric(TableObject$end)>end2] <- end2
TableObject$end[is.na(TableObject$end)] <- TableObject$start
GrangesObject <- makeGRangesFromDataFrame(TableObject, keep.extra.columns=TRUE)

if(as.numeric(TableObject$end[1])-as.numeric(TableObject$start[1]) > 100){
  showthis <- list(showFeatureId = T, feature = TableObject$variant.region.id, showId = F, stackHeight = 0.75, 
                   min.width = 1, rotation.item = 0)
} else {showthis <- list(showFeatureId = T, feature = TableObject$variant.region.id, showId = F, stackHeight = 0.2, 
                         min.width = 3.5, rotation.item = 35)}

levels(GrangesObject$type) <- c("#7AC5CD","#E6E6FA","#D8BFD8","#2B547E","#FAD1C1","#90EE90","#BFE3B6")
types <- as.character(GrangesObject$type)

```

### Create individual plot tracks

```{r tracks}
chr1 <- as.character(unique(seqnames(GrangesObject)))
itrack1 <- IdeogramTrack(genome = "hg19", chromosome = chr1,
                         background.panel = "#FFFEDB",
                         background.title = "#FFFEDB", cex = 1.2,
                         showBandId = TRUE, cex.bands = 0.85)
grtrack1 <- GeneRegionTrack(GeneModel, genome = "hg19", chromosome = chr1, 
                            name = "Gene Model",
                            transcriptAnnotation = "symbol",
                            background.title = "#EEC591",
                            background.panel = "#FFFEDB",
                            col.border.title = "dark gray",
                            cex.title = 1.1, showId = TRUE, cex.group = 0.9, 
                            col.id = "white")
gtrack1 <- GenomeAxisTrack(background.panel = "#FFFEDB",
                           background.title = "#FA8072",
                           range=IRanges(start = start,
                                         end = end,
                                         names = gene),
                           col.border.title = "dark gray", showId = TRUE, cex = 1.1)
atrack1 <- AnnotationTrack(GrangesObject,
                           name = typeofvariant,
                           background.title = "#7AC5CD",
                           background.panel = "#FFFEDB",
                           col.border.title = "dark gray",
                           cex.title = 1.1, 
                           feature = showthis$feature, 
                           cex.feature = 1,
                           fontcolor.feature = "#616771", cex.group = 0.5, stackHeight = showthis$stackHeight,
                           showFeatureId = showthis$showFeatureId, showId = showthis$showId, 
                           min.width = showthis$min.width, rotation.item = showthis$rotation.item,
                           fill = types)
```

### Plot tracks

Don't forget to show the plot in a new window.

```{r plot}
plotTracks(list(itrack1,gtrack1,grtrack1,atrack1),
           from = end1, to = end2, 
           col = NULL,
           add53 = TRUE, add35 = TRUE,
           fill.range = "#FA8072",
           fontface = 2, title.width = 0.3, 
           sizes = c(0.3,0.5,0.5,1.9),
           margin = 0, cex.id = 1.2)
```