---
title: "Gviz visualization of DPP8/9 genetic variation (hg19)"
output: html_document
---

### Fill in parameters

--- Before running this Rmd document, please fill in the following parameters: gene and typeofvariant ---

```{r para}
gene <- "Dipeptidyl Peptidase 9"
typeofvariant <- "Common short variants"

if(gene == "Dipeptidyl Peptidase 9"){start = 4675239 
                                    end = 4723855}
if(gene == "Dipeptidyl Peptidase 8"){start = 65734805
                                    end = 65810040}
end1 <- start-6240
end2 <- end+1380
```

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
```

```{r Rversion}
R.version$os
R.version.string
```

### Loading in GeneModel file

Make sure this R script is located in the correct subfolder where the NCBI results are.

```{r genemodel}
GeneModel <- read.csv(file="results-transcripts-UCSC-hg19.csv", header = TRUE, sep=";")
```

### Install and load Gviz Bioconductor package

```{r installation}
if (! requireNamespace("BiocManager", quietly = TRUE)){
   install.packages("BiocManager")}
if (! requireNamespace("Gviz", quietly = TRUE)) {
   BiocManager::install("Gviz")}
library(Gviz)
```

Bioconductor version:

```{r Bioconductor}
packageVersion("BiocManager")
```

### Check the version of the packages installed

```{r version}
packageVersion("Gviz")
packageVersion("S4Vectors")
packageVersion("stats4")
packageVersion("BiocGenerics")
packageVersion("parallel")
packageVersion("IRanges")
packageVersion("GenomicRanges")
packageVersion("GenomeInfoDb")
packageVersion("grid")
```

### Select the file containing the data to be visualized

```{r file}
filename <- read.csv(file=file.choose(), header = TRUE, sep = ";")
```

### Function to select the right columns from the table 

```{r function}
if (! requireNamespace("tidyr", quietly = TRUE)){
   install.packages("tidyr")}
library(tidyr)

Gviz_table <- function(filetable) {
  X <- filetable
 #################### GRCh37 coordinates in first assembly
  tablepart1 <- X[grep("^GRCh37", X$assembly1),]
  id1 <- tablepart1[1]
  if(any(names(tablepart1) == 'variation.type')){type1 <- tablepart1[3]} else {type1 <- tablepart1[5]}
  if(any(names(tablepart1) == 'variant.region.id')){regionid1 <- tablepart1[2]}
  if(any(names(tablepart1) == 'accession')){acc1 <- tablepart1[3]}
  if(any(names(tablepart1) == 'dbVar.variant.id')){clin1 <- tablepart1[10]}
  if(any(names(tablepart1) == 'ClinVar.variant.id')){clin1 <- tablepart1[4]}
  Chr_1a <- tablepart1[["Chr"]]
  Chr_1a <- gsub("[A-Za-z]", "", Chr_1a)
  Chr_1a <- sub("", "chr", Chr_1a)
  assembly1 <- tablepart1[["assembly1"]]
  assembly1 <- gsub(".*?:", "", assembly1)
  assembly1 <- separate(data = as.data.frame(assembly1), col = assembly1, into = c("start", "end"), sep = "-")
  if(any(names(tablepart1) == 'variant.region.id')) {
    GRCh37_assembly1 <- data.frame(id1,Chr_1a,assembly1,regionid1,type1,clin1)
  } else if(any(names(tablepart1) == 'accession')) {
    GRCh37_assembly1 <- data.frame(id1,Chr_1a,assembly1,acc1,type1,clin1)
  } else {
    GRCh37_assembly1 <- data.frame(id1,Chr_1a,assembly1,type1)
  }
  names(GRCh37_assembly1)[names(GRCh37_assembly1) == "assembly1"] <- "assembly"
  names(GRCh37_assembly1)[names(GRCh37_assembly1) == "Chr_1a"] <- "chr"
  #################### GRCh37 coordinates in second assembly
  tablepart2 <- X[grep("^GRCh37", X$assembly2),]
  id2 <- tablepart2[1]
  if(any(names(tablepart2) == 'variation.type')){type2 <- tablepart2[3]} else {type2 <- tablepart2[5]}
  if(any(names(tablepart2) == 'variant.region.id')){regionid2 <- tablepart2[2]}
  if(any(names(tablepart2) == 'accession')){acc2 <- tablepart2[3]}
  if(any(names(tablepart2) == 'dbVar.variant.id')){clin2 <- tablepart2[10]}
  if(any(names(tablepart2) == 'ClinVar.variant.id')){clin2 <- tablepart2[4]}
  Chr_1b <- tablepart2[["Chr"]]
  Chr_1b <- gsub("[A-Za-z]", "", Chr_1b)
  Chr_1b <- sub("", "chr", Chr_1b)
  assembly2 <- tablepart2[["assembly2"]]
  assembly2 <- gsub(".*?:", "", assembly2)
  assembly2 <- separate(data = as.data.frame(assembly2), col = assembly2, into = c("start", "end"), sep = "-")
  if(any(names(tablepart2) == 'variant.region.id') &
      any(names(tablepart2) == 'study.ID')){
    GRCh37_assembly2 <- data.frame(id2,Chr_1b,assembly2,regionid2,type2,clin2)
  } else if(any(names(tablepart2) == 'accession')) {
    GRCh37_assembly2 <- data.frame(id2,Chr_1b,assembly2,acc2,type2,clin2)
  } else {
    GRCh37_assembly2 <- data.frame(id2,Chr_1b,assembly2,type2)
  }
  names(GRCh37_assembly2)[names(GRCh37_assembly2) == "assembly2"] <- "assembly"
  names(GRCh37_assembly2)[names(GRCh37_assembly2) == "Chr_1b"] <- "chr"
  #################### GRCh37 coordinates in third assembly
  tablepart3 <- X[grep("^GRCh37", X$assembly3),]
  id3 <- tablepart3[1]
  if(any(names(tablepart3) == 'variation.type')){type3 <- tablepart3[3]} else {type3 <- tablepart3[5]}
  if(any(names(tablepart3) == 'variant.region.id')){regionid3 <- tablepart3[2]}
  if(any(names(tablepart3) == 'accession')){acc3 <- tablepart3[3]}
  if(any(names(tablepart3) == 'dbVar.variant.id')){clin3 <- tablepart3[10]}
  if(any(names(tablepart3) == 'ClinVar.variant.id')){clin3 <- tablepart3[4]}
  Chr_1c <- tablepart3[["Chr"]]
  Chr_1c <- gsub("[A-Za-z]", "", Chr_1c)
  Chr_1c <- sub("", "chr", Chr_1c)
  assembly3 <- tablepart3[["assembly3"]]
  assembly3 <- gsub(".*?:", "", assembly3)
  assembly3 <- separate(data = as.data.frame(assembly3), col = assembly3, into = c("start", "end"), sep = "-")
  if(any(names(tablepart3) == 'variant.region.id') &
      any(names(tablepart3) == 'study.ID')){
    GRCh37_assembly3 <- data.frame(id3,Chr_1c,assembly3,regionid3,type3,clin3)
  } else if(any(names(tablepart3) == 'accession')) {
    GRCh37_assembly3 <- data.frame(id3,Chr_1c,assembly3,acc3,type3,clin3)
  } else {
    GRCh37_assembly3 <- data.frame(id3,Chr_1c,assembly3,type3)
  }
  names(GRCh37_assembly3)[names(GRCh37_assembly3) == "assembly3"] <- "assembly"
  names(GRCh37_assembly3)[names(GRCh37_assembly3) == "Chr_1c"] <- "chr"
  ##### Gviz table
  GRCh37_Gviz <- rbind(GRCh37_assembly1,GRCh37_assembly2,GRCh37_assembly3)
  GRCh37_Gviz <- GRCh37_Gviz[order(-GRCh37_Gviz[1]),]
}
```

### Make GRanges Object from data frame and set colors to different variant (sub)types

```{r GRanges}
TableObject <- Gviz_table(filename)
TableObject
names(TableObject)[names(TableObject) == "variant.call.type"] <- "type"
names(TableObject)[names(TableObject) == "variation.type"] <- "type"
names(TableObject)[names(TableObject) == "clinical.assertion"] <- "interpretation"
names(TableObject)[names(TableObject) == "description"] <- "interpretation"
TableObject$start[as.numeric(TableObject$start)<end1] <- end1
TableObject$end[as.numeric(TableObject$end)>end2] <- end2
if(any(is.na(TableObject$end))== F) {
TableObject<-TableObject[!(as.numeric(TableObject$start)>end2 
                           | as.numeric(TableObject$end<end1)),]
}
if(gene == "Dipeptidyl Peptidase 9") {
TableObject$chr[TableObject$type == 'interchromosomal translocation'] <- "chr19"
} else {
TableObject$chr[TableObject$type == 'interchromosomal translocation'] <- "chr15"
}
TableObject$end[is.na(TableObject$end)] <- TableObject$start
#TableObject <- TableObject[which(TableObject$interpretation !="['Pathogenic']"),]
GrangesObject <- makeGRangesFromDataFrame(TableObject, keep.extra.columns=TRUE)


GrangesObject$type <- as.factor(gsub(".*//.*", "//",GrangesObject$type))
subtypes <- list("#FFB6C1"="copy number gain",
                          "#9ACD32"="copy number loss",
                                "#FFD700"="//",
                                "#FF00FF"="insertion",
                                "#FF00FF"="ins",
                                "#6db2c6"="inversion",
                                "#4575B4"="short tandem repeat",
                                "#BDB76B"="copy-neutral loss of heterozygosity",
                                "#FF0000"="interchromosomal translocation",
                                "#D2691E"="tandem duplication",
                                "#008080"="complex substitution",
                                "#C0C0C0"="copy number variation",
                                "#FF00FF"="novel sequence insertion",
                                "#00FFFF"="single nucleotide variant",
                                "#00FFFF"="snv",
                                "#98FB98"="delins",
                                "#DEB887"="Deletion","#DEB887"="deletion",
                                "#DEB887"="del",
                                "#D2691E"="Duplication","#D2691E"="duplication",
                                "#FF00FF"="alu insertion",
                                "#FF00FF"="sva insertion")
interpretlist <- list("#FF0000"="['Pathogenic']", 
                      "#708090"="['Uncertain significance']",
                       "#FF8C00"="['Likely pathogenic']",
                       "#008000"="['Benign']",
                       "#000000"="[]",
                       "#FF0000"="Pathogenic", 
                       "#708090"="Uncertain significance",
                        "#FF8C00"="Likely pathogenic",
                       "#008000"="Benign",
                        "#000000"="")

if(any(names(TableObject) == 'interpretation')){
  if(length(unique(GrangesObject$interpretation))==1) {
  levels(GrangesObject$interpretation) <- interpretlist
  clinint <- as.character(GrangesObject$interpretation)
  
  levels(GrangesObject$type) <- subtypes
  types <- as.character(GrangesObject$type)

  } else {
  pathogenic <- GrangesObject[which(GrangesObject$interpretation =="['Pathogenic']" 
                                    | GrangesObject$interpretation =="Pathogenic"),]
  uncertain <- GrangesObject[which(GrangesObject$interpretation =="['Uncertain significance']" 
                                    | GrangesObject$interpretation =="Uncertain significance"),]
  likelypath <- GrangesObject[which(GrangesObject$interpretation =="['Likely pathogenic']" 
                                    | GrangesObject$interpretation =="Likely pathogenic"),]
  benign <- GrangesObject[which(GrangesObject$interpretation =="['Benign']" 
                                    | GrangesObject$interpretation =="Benign"),]
  nointerpr <- GrangesObject[which(GrangesObject$interpretation =="[]" 
                                    | GrangesObject$interpretation ==""),]
  
  levels(pathogenic$type) <- subtypes
  types2 <- as.character(pathogenic$type)
  levels(uncertain$type) <- subtypes
  types3 <- as.character(uncertain$type)
  levels(likelypath$type) <- subtypes
  types4 <- as.character(likelypath$type)
  levels(benign$type) <- subtypes
  types5 <- as.character(benign$type)
  levels(nointerpr$type) <- subtypes
  types6 <- as.character(nointerpr$type)
  }
  
} else {
  clinint = "black"
  
  levels(GrangesObject$type) <- subtypes
  types <- as.character(GrangesObject$type)
}


GeneModel$symbol2 <- GeneModel$symbol
levels(GeneModel$symbol2) <- list("#202020"="NM_139159.5","#202020"="XM_011528408.1",
                                  "#202020"="XM_024451770.1","#202020"="XM_011528410.1",
                                  "#202020"="NM_130434.5","#202020"="NM_001320876.2",
                                  "#202020"="NM_001320875.2","#202020"="NM_197960.4")
GeneModel$symbol2 <- as.character(GeneModel$symbol2)
GeneModel$symbol2[is.na(GeneModel$symbol2)] <- "#A8A8A8"
fillmodel <- as.character(GeneModel$symbol2)


long <- GrangesObject[which(as.numeric(TableObject$end)-as.numeric(TableObject$start) > 100),]
short <- GrangesObject[which(as.numeric(TableObject$end)-as.numeric(TableObject$start) < 100),]

if(length(long) == 0 | length(short) == 0) {
if(as.numeric(TableObject$end[1])-as.numeric(TableObject$start[1]) > 100){
  showthis <- list(showFeatureId = T, feature = TableObject$variant.region.id, showId = F, 
                   stackHeight = 0.75, min.height = 3,
                   min.width = 1, rotation.item = 0)
} else {showthis <- list(showFeatureId = T, feature = TableObject$dbSNP.variant.id, showId = F, 
                         stackHeight = 0.2, min.height = 4,
                         min.width = 3.5, rotation.item = 35)}
} else {
  showlong <- list(showFeatureId = T, feature = long$variant.region.id, showId = F, 
                   stackHeight = 0.75, min.height = 3,
                   min.width = 1, rotation.item = 0)
  levels(long$type) <- subtypes
  typeslong <- as.character(long$type)
  
  showshort <- list(showFeatureId = T, feature = short$variant.region.id, showId = F, 
                   stackHeight = 0.2, min.height = 4,
                   min.width = 3.5, rotation.item = 35)
  levels(short$type) <- subtypes
  typesshort <- as.character(short$type)
}

```

### Create individual plot tracks

```{r tracks}
chr1 <- as.character(unique(seqnames(GrangesObject)))
itrack1 <- IdeogramTrack(genome = "hg19", chromosome = chr1,
                         background.panel = "#FFFEDB", lwd = 1.2,
                         background.title = "#FFFEDB", cex = 1.2,
                         showBandId = TRUE, cex.bands = 0.85, fontface = 2)
grtrack1 <- GeneRegionTrack(GeneModel, genome = "hg19", chromosome = chr1, 
                            name = "Gene Model",
                            transcriptAnnotation = "symbol",
                            background.title = "black",
                            background.panel = "#FFFEDB",
                            col.border.title = "black", col = NULL, fontface = 2,
                            cex.title = 1.1, showId = TRUE, cex.group = 0.9, fill = fillmodel)
gtrack1 <- GenomeAxisTrack(background.panel = "#FFFEDB",
                           background.title = "#FFFEDB", 
                           range=IRanges(start = start,
                                         end = end,
                                         names = gene),
                           col.border.title = "#FFFEDB", showId = TRUE, cex = 1.1,
                           col.range = "black", col = "black", lwd = 1, fontface = 2)

if(length(long) == 0 | length(short) == 0) {
if(any(names(TableObject) == 'interpretation') == F | length(unique(GrangesObject$interpretation))==1) {
atrack1 <- AnnotationTrack(GrangesObject,
                           name = typeofvariant,
                           background.title = "#A52A2A",
                           background.panel = "#FFFEDB",
                           col.border.title = "black",
                           cex.title = 1.1, 
                           feature = showthis$feature, 
                           cex.feature = 1,
                           fontcolor.feature = "#616771", cex.group = 0.5, 
                           stackHeight = showthis$stackHeight, min.height = showthis$min.height,
                           showFeatureId = showthis$showFeatureId, showId = showthis$showId, 
                           min.width = showthis$min.width, rotation.item = showthis$rotation.item,
                           fill = types, fontface = 2, fontcolor.feature = clinint, col = NULL)
} else {
atrack2 <- AnnotationTrack(pathogenic,
                           name = "!",
                           background.title = "#A52A2A",
                           background.panel = "#FFFEDB",
                           col.border.title = "black",
                           cex.title = 1.5, 
                           feature = pathogenic$variant.region.id, 
                           cex.feature = 1, min.height = showthis$min.height,
                           cex.group = 0.5, stackHeight = showthis$stackHeight,
                           showFeatureId = showthis$showFeatureId, showId = showthis$showId, 
                           min.width = showthis$min.width, rotation.item = showthis$rotation.item,
                           fill = types2, fontface = 2, fontcolor.feature = "#FF0000", col = NULL)
atrack3 <- AnnotationTrack(uncertain,
                           name = "?",
                           background.title = "#A52A2A",
                           background.panel = "#FFFEDB",
                           col.border.title = "black",
                           cex.title = 1.5, 
                           feature = uncertain$variant.region.id, 
                           cex.feature = 1, min.height = showthis$min.height,
                           cex.group = 0.5, stackHeight = showthis$stackHeight,
                           showFeatureId = showthis$showFeatureId, showId = showthis$showId, 
                           min.width = showthis$min.width, rotation.item = showthis$rotation.item,
                           fill = types3, fontface = 2, fontcolor.feature = "#708090", col = NULL)
atrack4 <- AnnotationTrack(likelypath,
                           name = "~!",
                           background.title = "#A52A2A",
                           background.panel = "#FFFEDB",
                           col.border.title = "black",
                           cex.title = 1.5, min.height = showthis$min.height,
                           feature = likelypath$variant.region.id, 
                           cex.feature = 1,
                           cex.group = 0.5, stackHeight = showthis$stackHeight,
                           showFeatureId = showthis$showFeatureId, showId = showthis$showId, 
                           min.width = showthis$min.width, rotation.item = showthis$rotation.item,
                           fill = types4, fontface = 2, fontcolor.feature = "#FF8C00", col = NULL)
atrack5 <- AnnotationTrack(benign,
                           name = "+",
                           background.title = "#A52A2A",
                           background.panel = "#FFFEDB",
                           col.border.title = "black",
                           cex.title = 1.5, 
                           feature = benign$variant.region.id, 
                           cex.feature = 1, min.height = showthis$min.height,
                           cex.group = 0.5, stackHeight = showthis$stackHeight,
                           showFeatureId = showthis$showFeatureId, showId = showthis$showId, 
                           min.width = showthis$min.width, rotation.item = showthis$rotation.item,
                           fill = types5, fontface = 2, fontcolor.feature = "#008000", col = NULL)
atrack6 <- AnnotationTrack(nointerpr,
                           name = "()",
                           background.title = "#A52A2A",
                           background.panel = "#FFFEDB",
                           col.border.title = "black",
                           cex.title = 1.2, 
                           feature = nointerpr$variant.region.id, 
                           cex.feature = 1, min.height = showthis$min.height,
                           cex.group = 0.5, stackHeight = showthis$stackHeight,
                           showFeatureId = showthis$showFeatureId, showId = showthis$showId, 
                           min.width = showthis$min.width, rotation.item = showthis$rotation.item,
                           fill = types6, fontface = 2, fontcolor.feature = "#000000", col = NULL)
}} else {
atrack7 <- AnnotationTrack(long,
                           name = "long",
                           background.title = "#A52A2A",
                           background.panel = "#FFFEDB",
                           col.border.title = "black",
                           cex.title = 1.1, 
                           feature = long$variant.region.id, 
                           cex.feature = 1, min.height = showlong$min.height,
                           cex.group = 0.5, stackHeight = showlong$stackHeight,
                           showFeatureId = showlong$showFeatureId, showId = showlong$showId, 
                           min.width = showlong$min.width, rotation.item = showlong$rotation.item,
                           fill = typeslong, fontface = 2, fontcolor.feature = "black", col = NULL)
atrack8 <- AnnotationTrack(short,
                           name = "short",
                           background.title = "#A52A2A",
                           background.panel = "#FFFEDB",
                           col.border.title = "black",
                           cex.title = 1.1, 
                           feature = short$variant.region.id, 
                           cex.feature = 1, min.height = showshort$min.height,
                           cex.group = 0.5, stackHeight = showshort$stackHeight,
                           showFeatureId = showshort$showFeatureId, showId = showshort$showId, 
                           min.width = showshort$min.width, rotation.item = showshort$rotation.item,
                           fill = typesshort, fontface = 2, fontcolor.feature = "black", col = NULL)
}
```

### Plot tracks

Don't forget to show the plot in a new window.

```{r plot}
if(length(long) == 0 | length(short) == 0) {
if(any(names(TableObject) == 'interpretation') == F | length(unique(GrangesObject$interpretation))==1) {
  plotTracks(list(itrack1,gtrack1,grtrack1,atrack1),
           from = end1, to = end2, 
           add53 = TRUE, add35 = TRUE,
           fill.range = "#A52A2A",
           title.width = 0.3, 
           sizes = c(0.3,0.5,0.5,1.9),
           margin = 0, cex.id = 1.2)
} else {
  a2 <- length(pathogenic$type)/10+0.05
  a3 <- length(uncertain$type)/10+0.05
  a4 <- length(likelypath$type)/10+0.05
  a5 <- length(benign$type)/10+0.05
  a6 <- length(nointerpr$type)/10+0.05
  plotTracks(list(itrack1,gtrack1,grtrack1,atrack2,atrack4,atrack3,atrack5,atrack6),
           from = end1, to = end2, 
           add53 = TRUE, add35 = TRUE,
           fill.range = "#A52A2A",
           title.width = 0.3, 
           sizes = c(0.3,0.3,0.3,a2,a3,a4,a5,a6),
           margin = 0, cex.id = 1.2)
}} else {
  a7 <- length(long$type)/10+0.05
  a8 <- length(short$type)/10+0.05
  plotTracks(list(itrack1,gtrack1,grtrack1,atrack7,atrack8),
           from = end1, to = end2, 
           add53 = TRUE, add35 = TRUE,
           fill.range = "#A52A2A",
           title.width = 0.3, 
           sizes = c(0.3,0.5,0.5,a7,a8),
           margin = 0, cex.id = 1.2)
}
```