---
title: "Gviz visualization of DPP8/9 genetic variation (hg19)"
output: html_document
---

### Fill in parameters

--- Before running this Rmd document, please fill in the following parameters: gene and resource (TCGA/ICGC/TumorPortal) ---

```{r para}
gene <- "Dipeptidyl Peptidase 9"
resource <- "TumorPortal"

if(gene == "Dipeptidyl Peptidase 9"){start = 4675239 
                                    end = 4723855}
if(gene == "Dipeptidyl Peptidase 8"){start = 65734805
                                    end = 65810040}
end1 <- start-6240
end2 <- end+1380
```

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
```

```{r Rversion}
R.version$os
R.version.string
```

### Loading in GeneModel file

Make sure this R script is located in the correct subfolder where the NCBI results are.

```{r genemodel}
GeneModel <- read.csv(file="results-transcripts-UCSC-hg19.csv", header = TRUE, sep=";")
```

### Install and load Gviz Bioconductor package

```{r installation}
if (! requireNamespace("BiocManager", quietly = TRUE)){
   install.packages("BiocManager")}
if (! requireNamespace("Gviz", quietly = TRUE)) {
   BiocManager::install("Gviz")}
library(Gviz)
```

Bioconductor version:

```{r Bioconductor}
packageVersion("BiocManager")
```

### Check the version of the packages installed

```{r version}
packageVersion("Gviz")
packageVersion("S4Vectors")
packageVersion("stats4")
packageVersion("BiocGenerics")
packageVersion("parallel")
packageVersion("IRanges")
packageVersion("GenomicRanges")
packageVersion("GenomeInfoDb")
packageVersion("grid")
```

### Select the file containing the data to be visualized

```{r file}
if (! requireNamespace("xlsx", quietly = TRUE)) {
   install.packages("xlsx")}
library(xlsx)

filename <- read.xlsx(file=file.choose(), sheetIndex = resource)
filename
```

### Make GRanges Object from data frame and set colors to different tumor types

```{r GRanges}
GrangesObject <- makeGRangesFromDataFrame(filename, keep.extra.columns=TRUE)

tumortypes <- list("#FBEA74"="Bladder urothelial carcinoma", "#FBEA74"="Bladder Urothelial carcinoma","#FBEA74"="BLCA",
                    "#C0C0C0"="Brain Lower Grade Glioma", "#C0C0C0"="Pediatric Brain Cancer",
                    "#b2d8d8"="Cervical squamous cell carcinoma and endocervical adenocarcinoma",
                    "#00008b"="Colon adenocarcinoma", "#00008b"="CARC", "#00008b"="CRC",
                    "#696969"="Glioblastoma multiforme", "#696969"="GBM",
                    "#800020"="Head and Neck squamous cell carcinoma", "#800020"="HNSC",
                    "#ff8c00"="Kidney renal clear cell carcinoma", "#ff8c00"="Renal Cell Cancer","#ff8c00"="KIRC",
                    "#FFFFFF"="Lung adenocarcinoma", "#FFFFFF"="LUAD",
                    "#e1c699"="Lung squamous cell carcinoma", "#e1c699"="Lung Cancer - Squamous cell carcinoma",  
                                                              "#e1c699"="LUSC",
                    "#000000"="Skin Cutaneous Melanoma", "#000000"="MEL",
                    "#996699"="Stomach adenocarcinoma",
                    "#ffe5b4"="Uterine Corpus Endometrial Carcinoma", "#ffe5b4"="UCEC",
                    "#ffc0cb"="Breast invasive carcinoma", "#ffc0cb"="Breast Triple Negative/Lobular Cancer", "#ffc0cb"="BRCA",
                    "#006666"="Ovarian serous cystadenocarcinoma", "#006666"="Ovarian Cancer", "#006666"="OV",
                    "#800080"="Pancreatic adenocarcinoma", "#800080"="Pancreatic Cancer", 
                                                           "#800080"="Pancreatic Cancer Endocrine neoplasms",
                    "#add8e6"="Prostate adenocarcinoma",
                    "#b2d8d8"="Thyroid carcinoma", "#b2d8d8"="Thyroid Cancer",
                    "#FFFF00"="Acute Myeloid Leukemia", "#FFFF00"="CLL",
                    "#CCCCFF"="Esophageal Adenocarcinoma", "#CCCCFF"="ESO",
                    "#50c878"="Liver Cancer",
                    "#ee82ee"="Malignant Lymphoma", "#ee82ee"="DLBCL",
                    "#ff0000"="Oral Cancer")
levels(GrangesObject$ttype) <- tumortypes
types <- as.character(GrangesObject$ttype)

GeneModel$symbol2 <- GeneModel$symbol
levels(GeneModel$symbol2) <- list("#202020"="NM_139159.5")
GeneModel$symbol2 <- as.character(GeneModel$symbol2)
GeneModel$symbol2[is.na(GeneModel$symbol2)] <- "#A8A8A8"
fillmodel <- as.character(GeneModel$symbol2)
```

### Create individual plot tracks

```{r tracks}
chr1 <- as.character(unique(seqnames(GrangesObject)))
itrack1 <- IdeogramTrack(genome = "hg19", chromosome = chr1,
                         background.panel = "#F5EEF8", lwd = 1.2,
                         background.title = "#F5EEF8", cex = 1.2,
                         showBandId = TRUE, cex.bands = 0.85, fontface = 2)
grtrack1 <- GeneRegionTrack(GeneModel, genome = "hg19", chromosome = chr1, 
                            name = "Gene Model",
                            transcriptAnnotation = "symbol",
                            background.title = "black",
                            background.panel = "#F5EEF8",
                            col.border.title = "black", col = NULL, fontface = 2,
                            cex.title = 1.1, showId = TRUE, cex.group = 0.9, fill = fillmodel)
gtrack1 <- GenomeAxisTrack(background.panel = "#F5EEF8",
                           background.title = "#F5EEF8", 
                           range=IRanges(start = start,
                                         end = end,
                                         names = gene),
                           col.border.title = "#F5EEF8", showId = TRUE, cex = 1.1,
                           col.range = "black", col = "black", lwd = 1, fontface = 2)
atrack1 <- AnnotationTrack(GrangesObject,
                           name = resource,
                           background.title = "#A52A2A",
                           background.panel = "#F5EEF8",
                           col.border.title = "black",
                           cex.title = 1.1, 
                           stackHeight = 0.2, min.height = 4,
                           showFeatureId = F, showId = F, 
                           min.width = 3.5, fill = types, 
                           fontface = 2, fontcolor.feature = "black", col = NULL)
```

### Plot tracks

Don't forget to show the plot in a new window.

```{r plot}
plotTracks(list(itrack1,gtrack1,grtrack1,atrack1),
           from = end1, to = end2, 
           add53 = TRUE, add35 = TRUE,
           fill.range = "#A52A2A",
           title.width = 0.3, 
           sizes = c(0.3,0.5,0.5,1.9),
           margin = 0, cex.id = 1.2)
```